services/elastic_ip_manager/src/main/java/com/futurewei/alcor/elasticipmanager/service/implement/ElasticIpServiceImpl.java

```diff
@@ -258,6 +269,56 @@ public ElasticIpInfo updateElasticIp(ElasticIpInfo request) throws Exception {
         return new ElasticIpInfo(eip);
     }
 
+    private Map<String, ElasticIp> getElasticIpsByPortId(String projectId, String portId) {
+
+        Map<String, Object[]> filter = new HashMap<>();
+        filter.put("project_id", new Object[] {projectId});
+        filter.put("port_id", new Object[] {portId});
+
+        return elasticIpRepo.findAllItems(filter);
+    }
+
+    private String getAssociatedPortIp(String projectId, String portId, Integer ipVersion, String ipAddress)
+            throws Exception {
+        PortManagerProxy portManagerProxy = new PortManagerProxy(projectId);
+        PortEntity port = portManagerProxy.getPortById(portId);
+
+        List<PortEntity.FixedIp> fixedIps = port.getFixedIps();
+        List<String> fixedIpv4List = new ArrayList<>();
+        for (PortEntity.FixedIp fixedIp: fixedIps) {
+            if (Ipv4AddrUtil.formatCheck(fixedIp.getIpAddress())) {
+                fixedIpv4List.add(fixedIp.getIpAddress());
+            }
+        }
+
+        if (fixedIpv4List.isEmpty()) {
+            throw new ElasticIpPipNotFound();
+        }
+
+        String associatedIp;
+        if (ipAddress != null) {
+            if (fixedIpv4List.contains(ipAddress)) {
+                associatedIp = ipAddress;
+            } else {
+                throw new ElasticIpPipNotFound();
+            }
+        } else {
+            if (fixedIpv4List.size() > 1) {
+                throw new ElasticIpMultipleFixedIpFound();
```

Comment:
I am trying to understand why port with multi-ips are invalid here. 

For a port with multiple private ips, one could be considered as primary ip and the rest are secondary. The EIP could be associated with the primary pip.

Reply:
That is reference from Neutron. Neutron will throw errors when associate an EIP with a multi addresses port and not specifing which address is actually associated with.  I could investigate further in primary ip in a port.

Reply:
Some references:

- AWS allows multiple private IPv4s assigned to a network interface, and each private IPv4 address can be associated with a single Elastic IP address, and vice versa.
Link: https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/MultipleIP.html#StepThreeEIP

- Azure allows similar configurations. Each VM may have multiple NICs, and every NIC attached to a VM has one or more IP configurations associated to it. Each configuration is assigned one static or dynamic private IP address. Each configuration may also have one public IP address (either dynamic or static) associated to it. 
Link: https://docs.microsoft.com/en-us/azure/virtual-network/virtual-network-multiple-ip-addresses-portal

Reply:
Thanks, that detailly references are very hopeful. For now we support each fixed ip to be associated with a indepandent EIP, but not recognize the primary ip of the fixed ips and could not automatically choose a fixed ip to be associated. Mightbe we could simply choose the first fixed ip when not specified, or we might think add a optional 'primary' attribute in fixed ip.

Reply:
Done. I removed the limitation

Reply:
The current implementation looks good.

Adding a new primary attribute to the fixed IP is ideal but requires more thinking as it involves change in multiple services. Let us discuss together with other team members and roll out a mid-term plan.